### PR TITLE
Add platform detection logic for SecureRandom

### PR DESCRIPTION
**Pull request by wilhuff**
Also, implement SecureRandom in terms of BoringSSL/OpenSSL, this makes the interface available on Windows and Android as well.

@paulb777 Is it okay to generate config.h this way in the podspec?

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/util/config.h.in [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+// This header defines macros for all available platform configuration values.
+// When building with CMake, it will substitute the lines marked with
+// cmakedefine with values corresponding to the local configuration.
+//
+// On Apple platforms we support building via CocoaPods without CMake. When
+// building this way we can't test the presence of features before building so
+// predefine all the platform-support feature macros to their expected values.
+
+#cmakedefine HAVE_ARC4RANDOM 1
```
Comment (paulb777) [inline Firestore/core/src/firebase/firestore/util/config.h.in]: why is the cmakedefine needed?  Won't COCOAPODS only be defined on Apple platforms?

Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/config.h.in]: The `#cmakedefine` allows cmake to substitute the correct values based on what the platform makes available.

These platform checks in `Firestore/core/src/firebase/firestore/util/CMakeLists.txt` drive the substitution of the values.

```
check_symbol_exists(arc4random stdlib.h HAVE_ARC4RANDOM)

# ...

check_include_files(openssl/rand.h HAVE_OPENSSL_RAND_H)
```

For example, on Windows and Android the `arc4random` function isn't available so we fall back on going to openssl.

File: Firestore/core/src/firebase/firestore/util/secure_random_openssl.cc [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#include <openssl/rand.h>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+SecureRandom::result_type SecureRandom::operator()() {
+  result_type result;
+  int rc = RAND_bytes(reinterpret_cast<uint8_t*>(&result), sizeof(result));
+  if (rc < 0) {
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/util/secure_random_openssl.cc]: OpenSSL claims 0 is also an error code. I think you want `if (rc <= 0)` or `if (rc != 1)`

https://www.openssl.org/docs/man1.1.0/crypto/RAND_bytes.html

Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/secure_random_openssl.cc]: Great catch. Thanks!

**Comment by wilhuff** [comment]
PTAL

**Comment by wilhuff** [comment]
Travis has refused to catch up; I've tested everything manually and this is good.

**Pull request merged**